### PR TITLE
feat(KONFLUX-4136): new reduce-snapshot script

### PR DESCRIPTION
**Pull request by scoheb**
- add new script to ec-cli image

### CODE REVIEW

File: hack/reduce-snapshot/reduce-snapshot.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      set +e
+      jq -e < "${SNAPSHOT}" > /dev/null 2>&1
+      CHECK_JSON_FILE=$?
+      set -e
+      if [ "${CHECK_JSON_FILE}" != "0" ]; then
+        echo "Cannot load snapshot from json string or file"
+        exit 1
+      fi
+      SNAPSHOT=$(cat "${SNAPSHOT}")
+    fi
```
Comment (zregvart) [inline hack/reduce-snapshot/reduce-snapshot.sh:36]: I think the `jq` commands here lack the filter. I think something simpler can be done here instead:

```suggestion
    SNAPSHOT="$(cat "${SNAPSHOT}" 2> /dev/null || echo "${SNAPSHOT}")"
    [[ ! "${SNAPSHOT}" =~ ^\s*\{ ]] && echo 'Error: Cannot load snapshot from JSON string or file' && exit 1
```

File: hack/reduce-snapshot/reduce-snapshot.sh [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    if [ "${COMPONENT_COUNT}" != "1" ] ; then
+      echo "Error: Reduced Snapshot has ${COMPONENT_COUNT} components. It should contain 1"
+      echo "       Verify that the Snapshot contains the built component: ${SNAPSHOT_CREATION_COMPONENT}"
+      exit 1
+    fi
+
+    echo "Reducing Snapshot to:"
+    echo "$REDUCED_SNAPSHOT" | jq .
+    SNAPSHOT=$(echo "$REDUCED_SNAPSHOT" | tr -d ' ' | tr -d '\n')
+    echo "$SNAPSHOT" > "${SNAPSHOT_PATH}"
```
Comment (zregvart) [inline hack/reduce-snapshot/reduce-snapshot.sh:52]: Might be dangerous if any of the keys or values in the JSON contain the space character, why not something like this:

```suggestion
    echo "$SNAPSHOT" | jq -c . > "${SNAPSHOT_PATH}"
```

Comment (scoheb) [reply hack/reduce-snapshot/reduce-snapshot.sh]: fixed. thanks

File: hack/reduce-snapshot/reduce-snapshot.sh [hunk 3]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env bash
+set -eux
```
[the remark this answers is not in the document]
Comment (scoheb) [reply hack/reduce-snapshot/reduce-snapshot.sh]: fixed. thanks

File: hack/reduce-snapshot/reduce-snapshot.sh [hunk 4]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (scoheb) [reply hack/reduce-snapshot/reduce-snapshot.sh]: done. thanks

File: hack/reduce-snapshot/reduce-snapshot.sh (hunk 2 above)
Comment (lcarva) [reply hack/reduce-snapshot/reduce-snapshot.sh]: I see this now:
```
echo "$REDUCED_SNAPSHOT" | jq .
SNAPSHOT=$(echo "$REDUCED_SNAPSHOT" | tr -d ' ' | tr -d '\n')
echo "$SNAPSHOT" | jq -c . > "${SNAPSHOT_PATH}"
```

I think @zregvart meant this:

```
# for displaying
echo "$REDUCED_SNAPSHOT" | jq .
# for saving
echo "$REDUCED_SNAPSHOT" | jq -c . > "${SNAPSHOT_PATH}"
```

Honestly, I would just do something like this:

```
echo "$REDUCED_SNAPSHOT" | jq '.' | tee "${SNAPSHOT_PATH}"
```

I don't really think we need to save the reduced snapshot as a compacted json.

Comment (scoheb) [reply hack/reduce-snapshot/reduce-snapshot.sh]: ok adjusted. but I seem to remember an issue whereby I needed to remove the line feeds...

**Review by lcarva — APPROVED** [review]

**Review by zregvart — APPROVED** [review]

File: Dockerfile [hunk 5]
```diff
@@ -72,6 +72,9 @@ COPY --from=build "/build/dist/ec_${TARGETOS}_${TARGETARCH}" /usr/local/bin/ec
 # Copy the one kubectl binary that can run in this container
 COPY --from=build "/build/dist/kubectl_${TARGETOS}_${TARGETARCH}" /usr/local/bin/kubectl
 
+# Copt reduce-snapshot script needed for single component mode
+COPY hack/reduce-snapshot.sh /usr/local/bin
```
Comment (simonbaird) [inline Dockerfile:76]: For tidiness I'd probably suggest:

```
COPY --from=build /build/hack/reduce-snapshot.sh /usr/local/bin
```

Comment (simonbaird) [reply Dockerfile:76]: (Not super important, also I realize I'm too slow and its merged already.)

**Pull request merged**